File: src/main/java/fi/metatavu/oioi/cm/rest/translate/WallDeviceApplicationTranslator.java
package fi.metatavu.oioi.cm.rest.translate;

import java.time.OffsetDateTime;
import java.util.Map;
import java.util.stream.Collectors;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import fi.metatavu.oioi.cm.persistence.model.Resource;
import fi.metatavu.oioi.cm.persistence.model.ResourceProperty;
import fi.metatavu.oioi.cm.persistence.model.ResourceStyle;
import fi.metatavu.oioi.cm.resources.ResourceController;
import fi.metatavu.oioi.cm.wall.WallDeviceApplication;

/**
 * Translator for wall device application
 * 
 * @author Antti Leppä
 * @author Heikki Kurhinen <heikki.kurhinen@metatavu.fi>
 */
@ApplicationScoped
public class WallDeviceApplicationTranslator extends AbstractTranslator<fi.metatavu.oioi.cm.persistence.model.Application, WallDeviceApplication> {
  
  @Inject
  private ResourceController resourceController;

  @Override
  public WallDeviceApplication translate(fi.metatavu.oioi.cm.persistence.model.Application entity) {
    if (entity == null) {
      return null;
    }

    WallDeviceApplication result = new WallDeviceApplication();
    result.setId(entity.getId());
    result.setName(entity.getName());

    Resource rootResource = entity.getRootResource();
    OffsetDateTime modifiedAt = entity.getModifiedAt();
    if (rootResource != null) {
      OffsetDateTime resourceModifiedAt = rootResource.getModifiedAt();
      if (resourceModifiedAt.isAfter(modifiedAt)) {
        modifiedAt = resourceModifiedAt;
      }
      result.setProperties(getProperties(rootResource));
      result.setStyles(getStyles(rootResource));
    }

    result.setModifiedAt(modifiedAt);
    return result;
  }

  /**
   * Translates styles
   * 
   * @param entity resource
   * @return styles as key value pairs
   */
  private Map<String, String> getStyles(Resource entity) {
    return resourceController.listStyles(entity).stream().collect(Collectors.toMap(ResourceStyle::getKey, ResourceStyle::getValue));
  }

  /**
   * Translates styles
   * 
   * @param entity resource
   * @return styles as key value pairs
   */
  private Map<String, String> getProperties(Resource entity) {
    return resourceController.listProperties(entity).stream().collect(Collectors.toMap(ResourceProperty::getKey, ResourceProperty::getValue));
  }
  
}

File: src/main/kotlin/fi/metatavu/oioi/cm/rest/translate/WallDeviceApplicationTranslator.kt
package fi.metatavu.oioi.cm.rest.translate

import fi.metatavu.oioi.cm.model.WallDeviceApplication
import fi.metatavu.oioi.cm.persistence.model.Application
import fi.metatavu.oioi.cm.persistence.model.Resource
import fi.metatavu.oioi.cm.resources.ResourceController
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Translator for wall device application
 *
 * @author Antti Leppä
 * @author Heikki Kurhinen <heikki.kurhinen></heikki.kurhinen>@metatavu.fi>
 */
@ApplicationScoped
class WallDeviceApplicationTranslator : AbstractTranslator<Application?, WallDeviceApplication?>() {

    @Inject
    lateinit var resourceController: ResourceController

    override fun translate(entity: Application?): WallDeviceApplication? {
        entity ?: return null

        val rootResource = entity.rootResource
        var modifiedAt = entity.modifiedAt
        var styles: Map<String, String>? = null
        var properties: Map<String, String>? = null

        if (rootResource != null) {
            val resourceModifiedAt = rootResource.modifiedAt!!
            if (resourceModifiedAt.isAfter(modifiedAt)) {
                modifiedAt = resourceModifiedAt
            }

            properties = getProperties(rootResource)
            styles = getStyles(rootResource)
        }

        return WallDeviceApplication(
            id = entity.id!!,
            name = entity.name!!,
            modifiedAt = modifiedAt!!,
            styles = styles ?: emptyMap(),
            properties = properties ?: emptyMap(),
        )
    }

    /**
     * Translates styles
     *
     * @param entity resource
     * @return styles as key value pairs
     */
    private fun getStyles(entity: Resource): Map<String, String> {
        return resourceController.listStyles(entity).associate { it.key!! to it.value!! }
    }

    /**
     * Translates styles
     *
     * @param entity resource
     * @return styles as key value pairs
     */
    private fun getProperties(entity: Resource): Map<String, String> {
        return resourceController.listProperties(entity).associate { it.key!! to it.value!! }
    }
}